### ISSUE TITLE
Sampling tricks

### ISSUE THREAD

**Issue by ordabayevy** [cellarium-ai/scvi-distributed#17]
we could use some of the webdataset tricks here: we can take a `buffer_size` parameter and shuffle indices within that buffer (rather than always shuffling within a single limits bin). If the dataset object can cache up to N chunks, we can allow ourselves to open as many chunks at once; but then again, sampling cells from multiple chunks means reading multiple chunks at once, which goes against the desire to reduce the data loading latency.

I think a cute idea could be to have a `linear_iterator` and then `__iter__`. Linear iterator just iterates through the list of cell indices in the order of appearance in the assigned chunks. The main `__iter__` takes the stream from the `linear_iterator` and modifies it:  it can buffer ahead and shuffle, it can `materialize` the next chunks halfway through the buffer. Once wrinkle to work out there is how to load the next chunks ahead of time without blocking the iterator; not sure how to do it (otherwise, there is no point is loading data ahead of time: if GPU cannot work while the data is being loaded, it doesn't matter whether we load all of the data at once or incrementally).

_Originally posted by @mbabadi in https://github.com/cellarium-ai/scvi-distributed/pull/16#discussion_r1114913743_

**Comment by mbabadi** [comment]
@fedorgrab Fedor, since you're an async guru (:D), do you know how this could be possibly done? In brief, could we read an AnnData file from GCS, open it, and put in a dictionary, without (or with minimal) blocking the interpreter? More concretely, let's say we'd like to iterate over a certain list and when we're halfway through the list, we want to load a big file on the background (and put it in a cache), but we don't want to wait for the load to complete before we yield the next item in the list.

**Comment by fedorgrab** [comment]
One of the possible solutions would be the following:
Create two processes (parallel executing tasks). One process does the iteration, while the other downloads the file. You can start them simultaneously (not sure if you really want to wait till the half of the loop). For that ProcessPoolExecutor would do the job:

```python3
import concurrent.futures as concurrency
import multiprocessing

def download_task(arg_1, arg_2):
    pass

def iteration_task(arg_1, arg_2):
    for i in range(big_number):
        pass


num_of_workers = multiprocessing.cpu_count()

with concurrency.ProcessPoolExecutor(max_workers=num_of_workers) as executor:
    download_task_future = executor.submit(download_task, arg_1="foo", arg_2="bar")
    iteration_task_future = executor.submit(iteration_task, arg_1="buzz", arg_2="foo")
    futures = [download_task_future, iteration_task_future]
    # You can wait till all of your tasks completed or till the first task completes
    # If smth went wrong, your not_done list will not be empty.
    done, not_done = concurrency.wait(futures, return_when=concurrency.ALL_COMPLETED)
```

**Comment by fedorgrab** [comment]
I am still confused with Threads and Processes. But according to this guide, for this case Thread is a better thing:
https://superfastpython.com/thread-long-running-background-task/

**Comment by mbabadi** [comment]
Great! _but_ we really want to trigger loading the file at specific times.  Here's the full context:  we want to iterate through cells from individual AnnData chunks in a pre-determined order (of chunks), e.g. chunk 1, chunk 3, chunk 5, etc.  We make a list of index of all cells contained in these genes, e.g.
```
1, 2, ..., 10_000, 30_001, 30_002, ..., 40_000, 50_001, 50_002, ..., 60_000
```
The `DistributedAnnDataCollection` takes care of loading the appropriate AnnData (given a cell index), caching it, etc.  Now, let's say the minibatch size is 100. After 50 minibatches, we're half-way through the first AnnData chunk. At this point, we want to signal the `DistributedAnnDataCollection` to load and cache the next chunk (here, chunk 5) while we go through the rest of chunk 1, so that when the time comes, chunk 5 is already cached and GPU doesn't have to wait for I/O.

**Comment by fedorgrab** [comment]
Would that work? 
```python3
import time
from concurrent import futures as concurrency


def downloading_task(foo):
    print(foo, "Downloading started")
    time.sleep(3)
    print("Downloading finished")


def task_1():
    with concurrency.ThreadPoolExecutor(max_workers=2) as executor:
        for i in range(10):
            print(i)
            if i == 3:
                _ = executor.submit(downloading_task, foo="YOOOO")

            time.sleep(2)


if __name__ == "__main__":
    task_1()

```
The output would be:
```
0
1
2
3
YOOOO Downloading started
4
Downloading finished
5
6
7
8
9
```

**Comment by fedorgrab** [comment]
If I understood your case, that should do the job. Please let me know if got it wrong.

**Comment by fedorgrab** [comment]
Also, you can submit your task, get a future, and then when your loop ends, you can
```python
concurrency.wait([future], return_when=concurrency.ALL_COMPLETED)
```

In this case, you will ensure that your task has finished, and if it didn't it will lock your execution and wait till downloading was done. You 100% sure your anndata was cached and only after that you can start new minibatch processing.

**Comment by mbabadi** [comment]
There's just one subtle issue: the distributed sampler lives on a different process than the `DisrtibutedAnnDataCollectionDataset`. I believe PyTorch's `DataLoader` makes as many as `num_worker` processes to allow parallel data loading. We want the distributed sampler to signal `DisrtibutedAnnDataCollectionDataset` in the other processes to load the next chunk, but that sounds too complicated. Perhaps we should let each `DisrtibutedAnnDataCollectionDataset` know about the order of chunks that it should load at the beginning of each epoch, and it would take care of pre-caching the next chunk (using your code snippet) .
